File: src/main/java/simplifying/method/calls/Notification.java
package simplifying.method.calls;

import java.util.List;

public class Notification {
    private static List<String> authorizedChannels = List.of("SMS", "EMAIL", "PUSH");
    private final String channel;

    private Notification(String channel) {
        this.channel = channel;
    }

    public static Notification createNotification(String channel) {
        if (channel == null
                || channel.isEmpty()
                || !authorizedChannels.contains(channel)) {
            throw new IllegalArgumentException("Invalid channel provided");
        }
        return new Notification(channel);
    }
}

File: src/scala/src/main/scala/simplifying/method/calls/Notification.scala
package simplifying.method.calls

import scala.util.{Failure, Success, Try}

final case class Notification private (channel: String)

object Notification {
  final val Sms = "SMS"
  final val Email = "EMAIL"
  final val Push = "PUSH"

  val authorizedChannels = List(Sms, Email, Push)

  def create(channel: String): Try[Notification] = {
    if (channel.isEmpty || !authorizedChannels.contains(channel))
      Failure(new IllegalArgumentException("Invalid channel provided"))
    else Success(Notification(channel))
  }
}
